### ISSUE TITLE
STM32F7 support

### ISSUE THREAD

**Issue by Untitled86** [NETMF/netmf-interpreter#482]
I've been using NETMF with STM32F4 chips for some time now.  Recently my employer started designing boards with STM32F7 chips.  Naturally I need to be able to support them, or ditch NETMF.  

Can someone point me at the best/simplest path to "port" NETMF to this new chip?  

I'm not an expert when it comes to MPUs.  I've done a lot of NETMF coding in C# and a fair amount of C/CPP customizations to our firmware/runtime (mostly tweaking hardware feature drivers and moving slow bits from C# to CPP).

Of course, I will post whatever I end up with if I'm able to get this working.

Thanks!

This probably goes without adding...  But NETMF is slipping into the abyss.  While embedded everything is exploding NETMF isn't.  Adding support for the latest STM chip (when it probably isn't a ton of work) seems like a good idea.  Unless it is a ton of work, in which case, please tell me so I can dump NETMF.

**Comment by LodinErikson** [comment]
Same here. We discussing to use the STM32F7-Disco as a base for the next development.
I read somewhere that the code for F4 should be the same for F7, but I don't know it.
I have many years of experience developing SMALL embedded systems, but a NETMF port is realy beyond my abilities.

**Comment by doingnz** [comment]
A port to STM 32F746GDISCOVERY specifically would be ideal as it is inexpensive (US $49), and includes LCD 480x272 with touch and various IO.

http://www.st.com/content/st_com/en/products/evaluation-tools/product-evaluation-tools/mcu-eval-tools/stm32-mcu-eval-tools/stm32-mcu-discovery-kits/32f746gdiscovery.html

**Comment by piwi1263** [comment]
Hi Guys, I just tried to use a 411 based image and tried it on a 746ZG doesn't fly, neither is it for the 412ZG. the latter one is even closer to the 411. Both giving me the USB Descriptor not found error. The code might be compatible but you need to rerun the compile of netmf with the appropriate dev specs. And there might be an adjustment here or there. Something alike https://github.com/NETMF/netmf-interpreter/blob/dev/Solutions/STM32F4DISCOVERY/platform_selector.h

**Comment by cw2** [comment]
I have not yet had a chance to dig deeper into particular differences between STM32F4 and F7, but in general, you'd most likely need to adjust interrupt controller configuration (usually, the number of IRQ handlers is different), memory layout (assuming F7 is superset of F4) and clock settings (in platform_selector.h mentioned above).

I have been already looking at F7 discovery board, but most likely will not get it before autumn. Unless, of  course, someone convinces me otherwise :)

**Comment by josesimoes** [comment]
Because it leverages ST's official HAL for any STM32 series, PR #434 can give you a good start on that, but it still requires some work before being completely usable...

**Comment by cw2** [comment]
Well, software is usually not a problem - embedded development boards have rather low [WAF](https://en.wikipedia.org/wiki/Wife_acceptance_factor), especially during the holiday season :cry:

**Comment by Untitled86** [comment]
Thanks for all the posts.  By the way, this amazing guy already did the port for Llilum.  Doesn't that make this port to NETMF simpler?  

https://github.com/roceh/stm32f7disco_llilum_lcdtest/issues/1

**Comment by smaillet-ms** [comment]
Ideally it would, however Llilum and NETMF use very different platform Abstraction layers at this point. One of the things I'm working on (and really need to commit to a complete written description) is to define a unified PAL that would allow both to share the same PAL allowing for better compare/contrast and migration as appropriate. 

That said, having working code as a reference for the startup and clocks, IRQs and Pin maps certainly makes things easier.

**Comment by Untitled86** [comment]
Today I'm attempting to add support to NETMF for F7.  I found the peripheral libraries online.   Starting with the F4 code I moved all the code and header files in place, renamed all the folders, updated all the project files, etc.  Next I updated the clock info in the Platform_selector.h from that other project.  Now I'm running into differences that are more substantial.  The bootstrap code for the F4 references a constant that doesn't exist in the F7 version of ST's H files.  ...Anyways

I'm going to see how far I can get on my own.  I'll upload my progress here and maybe someone can help me.  I have about a week to do this.  If I cannot make this happen I will likely be forced by my employer to switch to FreeRTOS.

This would be tragic.  I've done some pretty awesome things with NETMF.  One of our instruments (a UV-VIS-NIR spectrometer) uses complex math to detect how much of something is in the item being measured (chemometric models).  For example if you scan a mango it can tell you how much sugar is in it.  We implemented NIPLS (Non-iterative partial least squares) which works well but, because I used NETMF, I was able to allow users to write their own "plugins" using our desktop app (also C#) to implement their own math.   There are lots of various "maths" that work better for different purposes.  LDA for example is better at identifying things, but not as good at identifying the quantity of a given thing.  ...E.g. does the mango have any sugar vs how much sugar does it have.   I also gave users the ability to customize the screen and write macros (like Outlook rules) to change content dynamically.  ...Things I can't do with FreeRTOS.  This code is live, in production, being used successfully.  I am going to be very unhappy if I have to re-write this in C/CPP just because NETMF wasn't kept current.  ...I'm probably not the only one in this kind of situation.

If you're curious, this is that products page on our site (https://www.felixinstruments.com/food-science-instruments/portable-nir-analyzers/f-750-produce-quality-meter/).  Unfortunately our marketing people don't highlight the features I talked about.  Really they're more the type of feature that's important to have when you need it, but not really that sexy during a sale.  ...The instrument is so large because the spectrometer is a very expensive Zeiss MMS1 spectrometer.  That screen, it's a daylight visible memory LCD from Sharp (like E-Ink but better).  I have a nice light-weight SPI driver for that, which includes very fast drawing for fonts, etc.  The font characters are stored by their Unicode and the device even supports switching to Chinese/Hewbrew characters on the fly.  There's a small desktop app that converts the TrueType fonts to compressed black and white bitmaps in three font sizes.  And, most importantly, it renders them very quickly on the device.  ...The device performs great at 168Mhz with ~120K of internal memory, and 4MB of slow external memory.  The move to the F7 isn't for speed, we were motivated by the added support for cheaper faster external memory.

I've been too busy to share a lot of my code, but my employer was onboard with that the last time I spoke with him.  Anyways, rant complete.  The point is that I've invested too much in NETMF to give it up.

**Comment by piwi1263** [comment]
@Untitled86 There is light at the end of the tunnel, so do not abandon netmf on F7 to soon ...

**Comment by josesimoes** [comment]
@Untitled86 An awesome example of what can be done with NETMF!! Great work there. 👏

**Comment by LodinErikson** [comment]
**@Untitled86** Any news on this ?
Maybe I can give you (very) little help. I have no idea how to use the SDK, github or the needed folder structure, but I have a licensed MDK pro with some experience with it.
And yes, I know my English is terrible, but I try my best... or google translate :)

**Comment by Untitled86** [comment]
Yes!  I've got my board beeping and flashing LEDs!  No USB or other peripherals.  ...And I'm using the 4.3 codebase at the moment, just because I'm familiar with it.  I plan on using 4.4 once I sort out everything.

Some key important changes are the clock (of course) and memory address changes.  The F7 supports two flash modes,  ITCM or AXIM.  Those change the address you use to access flash memory.  It's 0x08000000 for AXIM and 0x00200000 for ITCM.  The boot option bytes control the mode used, along with how the board is wired.  Ours was wired for ITCM.  HOWEVER, in DFU mode (which is how I program boards) it ALWAYS boots to AXIM mode.  So confusingly you setup your firmware for 0x002... but when you use STDFU Tester you flash to 0x080... 

Also the SRAM + CCM addresses changed...
0x20000000 (64K CCM aka TCM)
0x20010000 (320K SRAM) 
0x2004C000 (16K SRAM)

I'm using these settings for my clock in platform_selector...

#define SYSTEM_CLOCK_HZ                 216000000
#define SYSTEM_CYCLE_CLOCK_HZ           216000000
#define SYSTEM_APB1_CLOCK_HZ            54000000
#define SYSTEM_APB2_CLOCK_HZ            108000000
#define SYSTEM_CRYSTAL_CLOCK_HZ         12000000   //Disco is 25MHz, we use 12Mhz

Now I just have to get all the peripherals working! :)  No problem right?  I'll post my code to my GitHub.

**Comment by Untitled86** [comment]
Ok, I've run into another wall.  I must be setting the clocks up wrong.  I used the F4's Bootstrap file as a starting place.  There the RCC_PLLCFGR_PLLN_BITS, RCC_PLLCFGR_PLLP_BITS, RCC_PLLCFGR_PLLQ_BITS, etc. are probably all being set wrong.  

I can execute this code no problem...
CPU_GPIO_EnableOutputPin((GPIO_PIN)((1 \* 16) + 12), true );
CPU_GPIO_EnableOutputPin((GPIO_PIN)((1 \* 16) + 12), false );

And my beeper beeps.  Well, it clicks.  with no delay it's barely audible.  If I don't put the pin low it beeps very loudly.  So that's working. But if I execute this code...
CPU_GPIO_EnableOutputPin((GPIO_PIN)((1 \* 16) + 12), true );
HAL_Time_Sleep_MicroSeconds(1000);
CPU_GPIO_EnableOutputPin((GPIO_PIN)((1 \* 16) + 12), false );

It hangs forever.  So...  If anyone wants to help, I've uploaded my bootstrap file to Onedrive...
https://1drv.ms/u/s!AsDrOapsdMSUk7ZsEG11OtNuqQzclQ

Also, any general advise if I'm headed in the wrong direction would be great.

Thanks in advance!  I'll be looking for examples of how other people setup the clock.  As mentioned above, I'm using a 12Mhz external clock instead of the 25Mhz used on the Disco.  And my target processor speed is 216Mhz.

**Comment by cw2** [comment]
Congratulations @Untitled86 on getting F7 working!

Full disclosure: I am working on NETMF port for NUCLEO-F746ZG board, which should arrive next week...

Regarding the problem with HAL_Time_Sleep_MicroSeconds hanging up, what compiler do you use? Have a look at generated assembly listing and ensure there are THUMB instructions for IDelayLoop function (in IDelayLoop.s) generated and verify jump addresses to have thumb bit set.

**Comment by LodinErikson** [comment]
**@Untitled86** Do you know "STM32CubeMX" software ? 
Maybe you can compare your settings with the auto generated code.

**Comment by SytechDesigns** [comment]
@cw2  and any one else interested... did you maybe want to collaborate on the M7 disco port, I do have some\devicecode.. code already : STM32F7_Bootstrap, STM32F7_flash, STM32F7_usart,STM32F7_usb and the processor selector.h etc - most of the other peripherals are compatible with the stmf4 ( I think spi will need some minor tweaks). I do need to check with the 'source' of the code whether it is ok to publically release them, at the moment they are for internal development.

**Comment by cw2** [comment]
@SytechDesigns if you have anything that you could share, it would be much appreciated. Although I am not currently focused on the Disco board (targeting simpler Nucleo), obviously the basic components are shared and it would save a lot of work; and also having working reference source code is very valuable.

Thank you in advance.

**Comment by LodinErikson** [comment]
**@SytechDesigns** : Sorry but I dont understand what you mean with "M7 disco".
Is this a different hardware then the "[STM32F746GDISCOVERY](http://www.st.com/content/st_com/en/products/evaluation-tools/product-evaluation-tools/mcu-eval-tools/stm32-mcu-eval-tools/stm32-mcu-discovery-kits/32f746gdiscovery.html)" and the "[NUCLEO-F746ZG](http://www.st.com/content/st_com/en/products/evaluation-tools/product-evaluation-tools/mcu-eval-tools/stm32-mcu-eval-tools/stm32-mcu-nucleo/nucleo-f746zg.html)" ?
Or the "[32F769IDISCOVERY](http://www.st.com/content/st_com/en/products/evaluation-tools/product-evaluation-tools/mcu-eval-tools/stm32-mcu-eval-tools/stm32-mcu-discovery-kits/32f769idiscovery.html)" ?
**@All** Maybe it is better doing a port wich is based on the auto generated files from the  "STM32CubeMX" software ?? I really don't know... 
So much to learn...

**Comment by piwi1263** [comment]
Hi Lutz, that's actually a marvelous idea, having STM32CubeMx generate the netmf port software as well. That would certainly boost the use of netmf ... the current process of porting is soo primitive ....

**Comment by LodinErikson** [comment]
Yesterday I downloaded all(?) files, puzzled about the paths and started to build the STM32f469IDisco and... 147errors :( 
I had hoped to find a Keil project by using the mdk setting, but nothing. Instead struggling with "msbuild" and 10 lines long commands and a 6MB logfile :( :( 
Maybe I should go back to my "small" controllers...

**Comment by cw2** [comment]
@LodinErikson Have you followed the [Getting Started](https://github.com/NETMF/netmf-interpreter/wiki/Getting%20Started) guide? I can build that solution successfully with

```
setenv_mdk 5.05 C:\Keil\ARM
msbuild Solutions\STM32F429IDISCOVERY\dotnetmf.proj
```

**Comment by LodinErikson** [comment]
**@cw2** Thanks for your help !
I had 2 problems.
1st : wrong path for CMSIS (not sure)
2nd : I updated Keil to newest Version V5.20 with C-Compiler Version V5.06.
With
`setenv_mdk 5.06 C:\Keil\ARM
`
I got 148 errors :(
With
`setenv_mdk 5.05 C:\Keil\ARM
`
everything is OK, no errors, no warnings :)
Someone else should check this. I am a beginner, maybe still my fault...

**Comment by cw2** [comment]
@LodinErikson Oh, I see - unfortunately, the version for MDK is a little bit confusing, the current value of '5.05' really means '5.05 and above' (until there are again breaking changes in the toolchain).

**Comment by Untitled86** [comment]
Yes, I'm using that for reference.  I never sorted out the HAL_Time_Sleep_MicroSeconds problem.  Switched to using HAL_Time_Sleep_MicroSeconds_InterruptEnabled and it worked perfectly?  I've been moving my beeping and LED flashing code around to see what works.  A lot of things ARE working properly.  USB isn't one of them.  I'm slowly going through the STM32F7_usb_functions.cpp file and comparing it with Cube and others.  Hopefully I'll sort it out soon.  This manual did mention that the USB wasn't SW compatible, so go figure.  http://www.st.com/content/ccc/resource/technical/document/application_note/73/76/21/47/ef/d0/4c/16/DM00164538.pdf/files/DM00164538.pdf/jcr:content/translations/en.DM00164538.pdf

**Comment by LodinErikson** [comment]
**@Untitled86** Any news on your project ?

**Comment by Untitled86** [comment]
@LodinErikson - Progress has been slow.  So far I haven't been able to get the USB feature working properly.

I don't know how to step through the firmware code running on my device, so I've been using my oscope and toggling some GPIOs to indicate what's happening.

Currently I can see the CPU_USB_Initialize function is running (in my STM32F7_usb_functions.cpp) on bootup.  It completes without error, but my USB device doesn't show up on the PC end.

Porting that from the F4 to the F7 required a lot of changes.  I used various working USB F7 projects for reference.  Most of them were simply renaming constants, but the structs also changed.  

My old F4 code had a single struct for the USB feature.  OTG_TypeDef.  The newer F7 header files have that broken into several structs.  USB_OTG_GlobalTypeDef, USB_OTG_DeviceTypeDef, USB_OTG_INEndpointTypeDef and USB_OTG_OUTEndpointTypeDef.

Endpoints in the old code were referenced using an array like this...
OTG->DIEP[i].INT = 0xFF; // clear endpoint interrupts
OTG->DOEP[i].INT = 0xFF;
OTG->DIEP[i].CTL = USB_OTG_DIEPCTL_EPDIS; // deactivate endpoint
OTG->DOEP[i].CTL = USB_OTG_DOEPCTL_EPDIS;

In the new code I modified that to use the new structs like this...
USB_OTG_IN_EP = ((USB_OTG_INEndpointTypeDef_) USB_OTG_FS_PERIPH_BASE + 0x900 + (i \* 0x20));
USB_OTG_OUT_EP = ((USB_OTG_OUTEndpointTypeDef_) USB_OTG_FS_PERIPH_BASE + 0xB00 + (i \* 0x20));

USB_OTG_IN_EP->DIEPINT = 0xFF; // clear endpoint interrupts
USB_OTG_OUT_EP->DOEPINT = 0xFF;
USB_OTG_IN_EP->DIEPCTL = USB_OTG_DIEPCTL_EPDIS; // deactivate endpoint
USB_OTG_OUT_EP->DOEPCTL = USB_OTG_DOEPCTL_EPDIS;

...I'm not sure if I made a mistake in modifying things, or what.  I'm really frustrated at this point. 

I also don't know if I'm initializing the clock properly.  I've tested my settings by toggling a GPIO at a specific speed (e.g. 1500ms).  I then monitored that GPIO using my scope and adjusted my RCC_PLLCFGR_PLLN_BITS, RCC_PLLCFGR_PLLP_BITS and RCC_PLLCFGR_PLLQ_BITS until I got exactly 1500ms on the scope.  ...But I wonder if the APB/AHB clock is being initialized improperly.

I'm over my head on this.  I'm not giving up, but this is really intense.  Any help is appreciated.

**Comment by cw2** [comment]
Well, I have TinyBooter and bare TinyCLR for STM32F7, with the obligatory LED blinky application working (USB deployment). There are a few things I have to finish before I can publish it, so please be patient - it will come 'soon' :)

**Comment by Untitled86** [comment]
Great news!  Thanks!

**Comment by LodinErikson** [comment]
@Untitled86 
I never tried USB-OTG but in the F7 demo was a bug in the USB-Host sources. 
Here is a post from the ST forum :

```
there was a post 

 "STMF327-DISCO and "MSC_Standalone" example don't work (Bug in USBH MSC lib v3.2.1)?"

 could not find it at the moment, search engine seems to have problems

 there is a bug in 
 Repository\STM32Cube_FW_F7_V1.1.0\Middlewares\ST\STM32_USB_Host_Library\Class\MSC\Src\usbh_msc.c

 line 446

 change 
 if((phost->Timer - MSC_Handle->timer) > 10000)
 to 
 if((phost->Timer - MSC_Handle->timer) < 10000)

```

( [https://my.st.com/df56f8d6] )

**Comment by Untitled86** [comment]
...Is anyone interested in earning $1,000+ helping me port the F7?  If anyone is seriously interested, and capable, my employer is OK with me contracting out some support on this problem.  LodinErikson sounds like he's close.  Maybe he'd be interested?  Of course, LodinErikson is using USB HOST and I need USB OTG (different pins and registers are involved).

**Comment by LodinErikson** [comment]
@Untitled86 
Oh, maybe there is some misunderstanding because of my limited English.
It was not me who found this bug in USB-Host.

If you need someone for simple 8051 or STM32 controllers based on Keil, yes I am in.
Or better, if you need someone for doing schematics and PCB layout (with Altium) I really can do that. :) 

But my skills are not good enough für your USB problem. :( 
Sorry.

**Comment by cw2** [comment]
@Untitled86 https://github.com/cw2/netmf-interpreter/tree/dev-stm32f746nucleo contains a bare STM32F746NUCLEO solution. It is still in an 'alpha' stage, the only working peripherals are GPIO and USB. I have verified GCC build, tinybooter (when active, all three user LEDs are on), application deployment and debugging from Visual Studio.

There are some things (marked with 'UNDONE' and/or 'FIXME' comments) that has to be finished and fixed, I'll most likely update the RAM memory layout in the next commit (move RAM-RW into SRAM2 if it fits, so whole SRAM1 is dedicated to heap). I am now working on serial support...

_Note: I am going to publish one or two PRs with fixes required for STM32F7 port separately (i.e. directly to the 'dev' branch), which means I'll most likely rebase dev-stm32f746nucleo branch._

Also, I would like to thank Peter and Justin for their support.

**Comment by cw2** [comment]
One important thing I forgot: The USB configuration uses the same VID/PID as other solutions, so the board could show with a wrong name, if there was another board attached previously. To fix that, uninstall the existing USB device in Device Manager and reattach the board.

There is some caching mechanism involved, the ideal/reliable solution is to use different PIDs for different boards or at least an unique serial number. If there is enough interest, I could implement automagic use of the CPU's unique device ID, but IIRC there were certain privacy concerns in the past.

**Comment by cw2** [comment]
@Untitled86 Regarding USB OTG, I would first need to have a look at it, no promises at the moment.

**Comment by Untitled86** [comment]
@cw2, Peter, Justin and others...

THANK YOU!  I have my custom F7 board working now.   Our EE configured our board/chip for ITCM instead of AXIM, and we have a 12Mhz external clock.  Those were the only things I had to account for with cw's code.  I'm now able to see my device via USB!

**Comment by Untitled86** [comment]
Adding support for SPI I ran into an interesting problem.  The new chip requires you cast the DR register as 8bits when you're attempting to send only 8bits via SPI.

On the F4...
spi->DR = out;

On the F7...
*(__IO uint8_t *)&spi->DR = out;

I'll post my code once I'm done testing everything else.  But this little issue tripped me up for a day.  ...I'm not the C/C++ expert I need to be.

**Comment by cw2** [comment]
@Untitled86  Interesting. Although personally I would probably do it the other way round, i.e. to extend the data to match the register size, so it is not accessed unaligned. But, I'd have to check the datasheet and reference code first...

**Comment by piwi1263** [comment]
I'm trying too to get SPI working for the F7 but can't even compile. SPI_CR1_DFF not declared in this scope. Anyone has an idea ?

**Comment by piwi1263** [comment]
Well found a workaround for the moment. Yet to see if SPI works, at least TinyClr is building, deploying.  I included in the STM32F7_SPI_functions.cpp:
![image](https://cloud.githubusercontent.com/assets/10466039/18024390/1c8e0e58-6c09-11e6-8d47-e5797e2e6cf3.png)

Resulting into reporting back 6 SPI ports.:
![image](https://cloud.githubusercontent.com/assets/10466039/18024397/33a722dc-6c09-11e6-80e9-ec19ecbcfa7b.png)

**Comment by Untitled86** [comment]
@piwi1263

My bad for not posting code.  I have the SPI working here... 

https://1drv.ms/u/s!AsDrOapsdMSUk7564f6E9jO_U_wynQ

There may be some features I'm missing.  I basically started with the F4 code, but my SPI display etc. are functioning properly with this code.

Note: I did NOT add proper support for all the F7 transfer size options.  The chip supports like 4-16bits.  The F4 based code I posted only supports 8 and 16bit transfer sizes.  And I've only tested the 8bit mode so far.

Also, specifically about SPI_CR1_DFF ...That bit in the CR1 register changed meanings.  They were using that to switch between 16 and 8bit mode in the F4 code.  Looking at the F4 header files SPI_CR1_DFF is defined as 0x0800 or (0000 1000 0000 0000 in binary).   According to the F7 pdf the 11th bit of that SPI_CR1 register is CRCL (CRC Length) bit of that register.  If you're using CRC this should be toggled to 1 or 0 if you're using 8 or 16bits.  

With the F7 SPI_CR2_DS controls the data size.  From 4-16bit.  I believe I set that to 0 which is an invalid value that causes the chip to default to 8bit mode.

**Comment by Untitled86** [comment]
Any idea what it takes to get the SD card working?  

I've added a folder solution\DeviceCode\BlockStorage\SD and populated it with SD_BL_Driver.cpp, stm32f7xx_hal_sd and stm32f7xx_ll_sdmmc (from STM examples).  ...Am I headed down the right path?

**Comment by piwi1263** [comment]
If you are looking into this on mbed you'll not only find how to support for native ili9341 based displays but how to use the SD card on the display as well. Scroll down until you see Display 3, that's exactly what I have ... [https://developer.mbed.org/users/dreschpe/code/SPI_TFT_ILI9341/wiki/Homepage](https://developer.mbed.org/users/dreschpe/code/SPI_TFT_ILI9341/wiki/Homepage)

**Comment by Untitled86** [comment]
@piwi1263 

Thanks but I'm using a custom board we made with a 2" Sharp Memory LCD.  I already had a driver for my display that I wrote back when I was rocking the F4. 

I think my problem with the SD card is getting the F7 example code to plugin to the NETMF codebase properly.  There are tiny annoying differences.  Like the F7 example code reads and writes in 32bit arrays but NETMF uses 8bit arrays.  Naturally I'm casting those, and doing everything else I think I need to be doing, but clearly I'm missing something.

**Comment by Untitled86** [comment]
@Everybody

SD card working now :) 

https://1drv.ms/u/s!AsDrOapsdMSUk757Q3ja_zfpegyoug

You'll also have to modify your TinyCLR, etc.  I'm just posting the driver files for now.  Hopefully I'll get my entire codebase up soon.

**Comment by piwi1263** [comment]
@Untitled86 Too bad your SPI Driver didn't work for me and my ILI9341 Display .... yet

**Comment by Untitled86** [comment]
@piwi1263 

My bad.  I posted the wrong code.  One sec.

**Comment by Untitled86** [comment]
@piwi1263 

Sorry about that.  Here's the code I'm using now.  I just tested it again and it works great for me.  This should work with 16 or 8bits.  My display is 8bit so I haven't tested 16bit with a device, but I've seen the output on my scope and it looks right.

Also note, the speed is fixed.  Just update the logic around CR1_BR if that's important to you.  Soon I'll have that working properly as well.

https://1drv.ms/u/s!AsDrOapsdMSUk758w9KoEEJ51jAa4w

**Comment by piwi1263** [comment]
@Untitled86 - That does something but only with some patience and a bit of luck and than ....

![image](https://cloud.githubusercontent.com/assets/10466039/18321028/95e9d81e-752c-11e6-8f19-05366fc154d1.png)

Updated with your driver, deployed TinyClr, but didn't play at all no flashing LEDs as before and no output either just waiting in the VS deploy window ... then after some time like 30 secs, the display shows up as above and stays for 2 secs and goes blank again ... 

Some strange effect is happening, what clock frequency are you running .. ?

Oh, it's time for the fruit show ....

**Comment by piwi1263** [comment]
[Update] - now after some minutes reaching the first breakpoint it shows the LCD screen permanently and after continuing the program, the LEDs are flashing again and the LCD stays ... strange behavior ... I guess that's what you get balancing on the bleeding edge ... 

[Update2] - After a reset it takes about 3 minutes and 45 seconds to have the display on and before the LEDs are flashing, based upon 216 MHz clock frequency. 

[Update3] - And running @192MHz it is pretty much the same time, almost to the second.

**Comment by Untitled86** [comment]
@piwi1263 - I'm running at 192Mhz.  And the SPI is super slow right now.  If you update the CR1_BR logic I'd appreciate the fix.  I've moved on to my external SDRAM.  Again, using a custom board so I can't just drop in example code. :/  I'll post the SDRAM code when that's working just incase anyone else needs it.

**Comment by piwi1263** [comment]
I'll have a look at the CR1_BR and try changing a bit here or there ... but no promises on the positive outcome of that ... :smile:

**Comment by cw2** [comment]
Regarding the SPI DR access, I was wrong: 8-bit data transfer requires that the DR register is read/written via 8-bit access, otherwise data will be automagically packed into 16 bits, which will cause problems with events in case of odd sequence of frames.

**Comment by Untitled86** [comment]
@piwi1263 

I just got my SDRAM working :)  Next is I2C and AnalogInputs then I'll circle back and improve the SPI CR1_BR logic if you haven't by then.  I'm not the world's greatest bit-shifter.  I have to slowly think it through with a calculator.

I'm using an inexpensive AS4C16M16S-6TCNTR chip.  256MB of 166MHz SDRAM for ~$2.  If anyone cares I can release that code now.  Otherwise I'll hold off until I've made all my changes to publish the entire repo.

**Comment by piwi1263** [comment]
My rather not so successful trails on the SPI so far:
- Altering the events of things, like setting CR1, CR2, GPIO etc. -> Mostly freezes up the board to the point one has to erase all memory and do a complete TinyBooter+TinyClr burn.
- Setting the prescaler to 2 effectively does the same, no response, no ping from mfdeploy. -> Require an erase and a complete burn again.
- At the moment it is running @216 with a hardcoded prescaler of SPI_BAUDRATEPRESCALER_64

I've switch over to SPI4 to rule out any influence from to network PHY (it is using PIN A7 too). Found that out using CubeMx and @cw2 confirmed the finding, but no change, the display starts only after 3'45" no matter what I do. But it displays and keeps on being stable.

For as far as I can see (I'm not the netmf expert, so correct me if I'm wrong), the SPI code has room for improvement, initiated/deinitiating SPI at the transaction level for all cases even if the bit size and other settings are not changed at all does not speed up the process if you got a clean/pure 8bit or whatever bit size SPI usage or only perform a deinit/init if the bit size or the settings are changing. 

Another question is what is the purpose of the delay for almost every transaction of 3 seconds (if I understand it correctly) and what is the hardcoded pin (16+12) used for ? And why is it only in the 8bit transaction and not in the 16bit one ? 

![image](https://cloud.githubusercontent.com/assets/10466039/18376064/dce331ae-765b-11e6-8377-58bf32dfdf86.png)

**Comment by piwi1263** [comment]
{Update] - Removed those delays and SPI displays the few string now within 7 secs. where the driver init and the LCD clear do only take appx 250 ms and 243 ms. Where a 3X draw string takes almost 4 secs. This all running @216 MHz clock with SPI_BAUDRATEPRESCALER_16.

Back to the bit shifting with the clock & prescaler ...

**Comment by Untitled86** [comment]
@piwi1263  - :)  ...Sorry.  I have a beeper on B12 that's what the delay is for.  I left in the debug code.  Also, I'm still running at 192MHz.  Do you have your devenv setup so you can step through native runtime code running on the device?  I don't.  So I use my scope and the beeper to see what's going on.  ...I've been coding in Textpad and compiling with GCC 5.4

**Comment by Untitled86** [comment]
I'm back on the SPI speed issue now that my memory is working :)   ...I'm pretty confused myself.  I've got a scope setup and I'm expecting the period to change on my SPI clock when I modify the CR1_BR space.  That doesn't appear to be happening.

I'd like to use something OTHER than my scope and the beeper to debug this.   In the past I've used statements like this in the native code to at least print back details...

`CLR_Debug::Printf("cr1: %d", (int)cr1);`
`CLR_Debug::Flush();`

Unfortunately, that isn't working for me.  I'm confused as to why this is.  My Platform_Selector.H has these values...

 `#define DEBUG_TEXT_PORT                 USB1`
`#define STDIO                   USB1`
`#define DEBUGGER_PORT                   USB1`
`#define MESSAGING_PORT                  USB1`

...I thought that meant my CLR_Debug.Printf's would go to the USB and ultimately Visual Studio's output window. 

~~If anyone can help me sort this out it'll simplify all my future work on this F7 business.~~ 

...Such a rookie mistake.  I was adding those debug statements to STM32F4_SPI_functions.cpp instead of STM32F7_SPI_functions.cpp.  ...Had both open for reference.  I'm leaving this comment here incase anyone else makes my stupid mistake.  It's working properly now.  If anyone else needs to do this be sure to add `#include <TinyCLR_Runtime.h>` at the top of your cpp file.

**Comment by cw2** [comment]
@Untitled86  Have you tried `hal_printf()` instead?

Personally, I tend to use separate transports for DebugTextPort and STDIO, such as serial (COMx) or ITM (ITM0), so debug/tracing messages do not alter timing too much.

Does your board have any debugging support? Such as Serial Wire Debug or JTAG? If yes, I would highly recommend you to get a hardware debugger - you can either use an on-board ST-LINK from Nucleo or Discovery board or ST-LINK/V2 ($20). It pays for itself almost instantly (especially if it's a commercial project).

**Comment by Untitled86** [comment]
@piwi1263

So, I simply uncommented the code that was in there for the F4 pertaining to the CR1_BR space and checked everything out.  It _sortof_ works.  Here's what I did...

1) I wrote tiny C# app that increases the SPI speed in a loop.
2) I connected my scope and setup the trigger to measure the period of the SPI clock.
3) I added those CLR_Debug.Printf lines to spit out the CR1_BR values.

![img_1219](https://cloud.githubusercontent.com/assets/1141713/18452890/a0105e80-78f1-11e6-80d7-ee53261c5741.JPG)
![snibitfordfotnet](https://cloud.githubusercontent.com/assets/1141713/18452901/a588e2e2-78f1-11e6-8fc3-e01235feaa3c.PNG)
![captu2re](https://cloud.githubusercontent.com/assets/1141713/18452905/a9603cee-78f1-11e6-859e-b4a5cd435417.PNG)

As you can see the speed is changing.  But it's not using all the possible register values.  It's only using 111, 101, 011 or 001.  No 110, 000, 010, 100.  So next we fix the logic.  I just wanted to post this so you (and others) could see my simple process.

...My scope is cheap, don't laugh, I'm poor.  And if you're wondering why that isn't a square wave, there's a resistor inline to prevent the scope's capacitance etc. from messing up my board's signals.  ...In reality it should be square.

EDIT: I'm glad I posted this but I believe I was reading the CR1 incorrectly.  I was using bits 2-4 for CR1_BR not bits 3-5.  See m later post.  I ended up writing more readable code but I believe the original code may have been fine.

**Comment by Untitled86** [comment]
@cw2 

Thanks.  I have a hardware debugger and JTAG pins on the board.  The guy who sits next to me codes RTOS all day.  Naturally he uses the JTAG.  He probably could help, but he doesn't like to share and isn't a team player.  Is there a doc online that walks me through setting up my netmf environment for JTAG debugging?

**Comment by piwi1263** [comment]
@cw2, @Untitled86 - I've always wondered where these printf, hal_printf were going. I've read something about a driver on mbed that can claim that output, but I haven't tried that. For native code output, debugging etc. I'm totally blind. I let it run, either it does work or if not, I'm tinkering til it works by trail and error or try a look alike native code in VisualGDB and step thru that, that'll run in VS and show register values, does step by step etc. But that just doesn't fly against the native code executed by netmf etc.

Has anyone tried this solution [http://www.carminenoviello.com/2015/03/02/how-to-use-stm32-nucleo-serial-port/] I do get the virtual com port but doesn't spit out anything.

I second that on the doc request !!!

**Comment by Untitled86** [comment]
@piwi1263 

I haven't done the virtual serial port thing.  About the doc, I'll write one if/when I figure this out.

Also, I'm not so sure about the original code.  I got confused about the max speed of SPI2 on the F7.  Anyways, I'm now using this code which is much more readable...

```
    if (Configuration.Clock_RateKHz <= (clock / 256))
    {
        cr1 |= SPI_CR1_BR_2 | SPI_CR1_BR_1 | SPI_CR1_BR_0;
    }
    else if (Configuration.Clock_RateKHz <= (clock / 128))
    {
        cr1 |= SPI_CR1_BR_2 | SPI_CR1_BR_1;
    }
    else if (Configuration.Clock_RateKHz <= (clock / 64))
    {
        cr1 |= SPI_CR1_BR_2 | SPI_CR1_BR_0;
    }
    else if (Configuration.Clock_RateKHz <= (clock / 32))
    {
        cr1 |= SPI_CR1_BR_2;
    }
    else if (Configuration.Clock_RateKHz <= (clock / 16))
    {
        cr1 |= SPI_CR1_BR_1 | SPI_CR1_BR_0;
    }
    else if (Configuration.Clock_RateKHz <= (clock / 8))
    {
        cr1 |= SPI_CR1_BR_1;
    }
    else if (Configuration.Clock_RateKHz <= (clock / 4))
    {
        cr1 |= SPI_CR1_BR_0;
    }
    else
    {
        //Leave the BR values at 0, that's the fastest.
    }

```

And these are my test results...

```
original_cr1: 836 | updated_cr1: 892 | Configuration.Clock_RateKHz: 93 | clock: 24000 | cr1_br 0-2: 111
original_cr1: 836 | updated_cr1: 884 | Configuration.Clock_RateKHz: 187 | clock: 24000 | cr1_br 0-2: 110
original_cr1: 836 | updated_cr1: 876 | Configuration.Clock_RateKHz: 375 | clock: 24000 | cr1_br 0-2: 101
original_cr1: 836 | updated_cr1: 868 | Configuration.Clock_RateKHz: 750 | clock: 24000 | cr1_br 0-2: 100 
original_cr1: 836 | updated_cr1: 860 | Configuration.Clock_RateKHz: 1500 | clock: 24000 | cr1_br 0-2: 011  
original_cr1: 836 | updated_cr1: 852 | Configuration.Clock_RateKHz: 3000 | clock: 24000 | cr1_br 0-2: 010  
original_cr1: 836 | updated_cr1: 844 | Configuration.Clock_RateKHz: 6000 | clock: 24000 | cr1_br 0-2: 001
original_cr1: 836 | updated_cr1: 836 | Configuration.Clock_RateKHz: 12000 | clock: 24000 | cr1_br 0-2: 000  
original_cr1: 836 | updated_cr1: 836 | Configuration.Clock_RateKHz: 24000 | clock: 24000 | cr1_br 0-2: 000  
original_cr1: 836 | updated_cr1: 836 | Configuration.Clock_RateKHz: 48000 | clock: 24000 | cr1_br 0-2: 000 
```

...So, everything is working as expected for me now.  I suggest you copy&paste my logic.  The other code was much faster, but I don't really care about an extra ms or two while configuring SPI.  I really only care about my transfer rates.

**Comment by Untitled86** [comment]
So I started during the I2C ... There are a lot of changes.  The I2C feature is interrupt driven, and the interrupt specific registers have moved AND have been split up.  I only need the I2C for one thing, driving and DAC for a speaker that beeps.  ...And I'm only using a square wave (seems to be the most annoying beep).  Bottom line, I'm using GHI's SoftwareI2C and skipping the hardware feature for now. 

I moved on to the Analog hardware feature.  That was trivial, it's already done and tested.  DAC and the WatchDog are up tomorrow.   

My latest beta Analog Input and SPI code.  Be sure to uncomment those features in TinyClr.Proj
https://1drv.ms/u/s!AsDrOapsdMSUk75-zRQN_Mbz33WCiw

**Comment by cw2** [comment]
Regarding diagnostic output, I have added a little doc in my repo: [Diagnostics and Troubleshooting](https://github.com/cw2/netmf-interpreter/tree/dev-stm32f746nucleo/Solutions/STM32F746NUCLEO#diagnostics-and-troubleshooting).

**Comment by cw2** [comment]
@Untitled86 You've got very nice setup there! But, do yourself a favor and get [ST-LINK/V2](http://www.digikey.com/product-detail/en/stmicroelectronics/ST-LINK-V2/497-10484-ND/2214535) - it's one of the most affordable units and it basically pays for itself instantly (no matter if your project is a commercial or hobby one, the productivity is increased by several orders of magnitude).

**Comment by cw2** [comment]
I have added the following topics to the [Getting Started](https://github.com/NETMF/netmf-interpreter/wiki/Getting%20Started) wiki page:
- [Flashing STM32Fx firmware](https://github.com/NETMF/netmf-interpreter/wiki/Flashing-STM32Fx-firmware)
- [Diagnostics and Troubleshooting](https://github.com/NETMF/netmf-interpreter/wiki/Diagnostics-and-Troubleshooting)

_I'll include the description of serial (COM) transport and how to view it in MFDeploy in the next update._

Any feedback, corrections, extensions etc. are welcome.

HTH

**Comment by Untitled86** [comment]
@cw2 

I have purchased an ST-LINK/V2 I just haven't taken the time to figure out how to use it to debug code.  Can I use Visual Studio?  If not, Eclipse?  What's the IDE of choice here?

For now I'll just use the Serial Wire Viewer you linked to in your Diagnostics and Troubleshooting.  That's at least a start.

**Comment by Untitled86** [comment]
Wow...  Ok, I'm skipping the Watchdog implementation as well.  I was using this managed code for my F4 board and I like how simple it is.  For anyone who's interested...

https://1drv.ms/u/s!AsDrOapsdMSUk78C25jk6mOK2JFqiw

This works on F4 and F7.  Since no change was required, it doesn't make sense for me to put any effort into this feature.  I'm moving on to the DAC now.

**Comment by cw2** [comment]
@Untitled86 I think you cannot use Visual Studio directly, you'd probably need [VisualGDB extension](http://visualgdb.com/), which I am afraid is not available for free.

Visual Studio Code supports debugging C++ targets via GDB and OpenOCD provides compliant interface, so it could work in a similar way like VisualGDB, but I have not tried that yet.

There are [numerous IDEs](https://en.wikipedia.org/wiki/List_of_ARM_Cortex-M_development_tools) for ARM embedded development, for Eclipse-based there is for example [Tutorial: Using Eclipse + ST-LINK/v2 + OpenOCD to debug](https://community.particle.io/t/tutorial-using-eclipse-st-link-v2-openocd-to-debug/10042).

The STM32 ST-LINK utility can read registers and do single instruction steps, on the **Target** menu select **MCU Core...**. But that is really very basic.

**Comment by Untitled86** [comment]
Analog Output is working now.  Here's the pre-release if anyone cares.  Again, don't forget to add this feature to your TinyClr.Proj as follows....

```
  <ItemGroup>
    <RequiredProjects Include="$(SPOCLIENT)\DeviceCode\Targets\Native\STM32F7\DeviceCode\STM32F7_DA\dotNetMF.proj" />
    <DriverLibs Include="STM32F7_DA.$(LIB_EXT)" />
  </ItemGroup>
```

https://1drv.ms/u/s!AsDrOapsdMSUk78DP6VtlA-_gSa9TA

**Comment by smaillet-ms** [comment]
Actually you could use VisualStudio if you want. It supports GDB based engines, we have support for that in the Llilum repo. Llilum also includes a C++ project system that can build C++ code as well, albeit rather clumsily at the moment. Improving that and sharing the system and VS support between NETMF and Llilum is on the radar for vNext.

**Comment by cw2** [comment]
If I understand it correctly Llium uses the same (MIEngine) support as [Visual C++ for IoT Development](https://visualstudiogallery.msdn.microsoft.com/35dbae07-8c1a-4f9d-94b7-bac16cad9c01) extension (?)

I have installed this extension and with the latest OpenOCD dev version (required for STM32F7) I was able to perform some debugging on the Nucleo-144 board. But, it is very unstable and I think the NETMF memory layout has to be changed to provide space for OpenOCD work area.

It looks very promising.

**Comment by smaillet-ms** [comment]
@cw2 Yes, under the hood it is the same MIDebugEngine - we have been using it with PyOCD for Llilum but there is some support for OpenOCD as well but that wasn't used or tested heavily by the team. VsCode ultimately uses the same engine for GDB support as well. The project system for C++code in Llilum is based on very early work from the VS team that has progressed significantly since. So the goal is to loop back around on all of this common tooling for Micro Controller scenarios and unify it so it is useable for NETMF, Llilum or just plain C/C++ based systems. (I waffle back and forth on whether that warrants setting it up as it's own independent project in GitHub that is then used by the others... Thoughts from the community and hopeful future development team is welcome on that.  )

**Comment by Untitled86** [comment]
I've found a problem with the SPI code I posted.  The code failed to properly send/receive data when I specified a speed over ~8MHz.  I seem to have solved the problem by using HAL_GPIO_Init from the ST libs.  I suspect half the issue was caused by the fact that I'm using non-standard pins for SPI1.  Normally it'd be PC1, PC2 and PC10.  I'm using PB13, PB14 and PB15.  But I also think the speed of the GPIO wasn't being set properly.  I'll post again when I know more.

I'm not posting code yet because I want to polish it up a bit, but if anyone else has trouble consider adding something like this...

`

switch (Configuration.SPI_mod)
    {    

```
//...Other cases here but omitted

case 1:
    RCC->APB1ENR |= RCC_APB1ENR_SPI2EN;

    GPIO_InitStructure.Pin = GPIO_PIN_13 | GPIO_PIN_14 | GPIO_PIN_15; 
    GPIO_InitStructure.Mode = ((uint32_t)0x00000002U);//GPIO_MODE_AF_PP 
    GPIO_InitStructure.Pull = ((uint32_t)0x00000000U); //GPIO_NOPULL
    GPIO_InitStructure.Speed = 0x00000003U; //GPIO_SPEED_FREQ_VERY_HIGH
    GPIO_InitStructure.Alternate = GPIO_AF5_SPI2;
    HAL_GPIO_Init(GPIOB, &GPIO_InitStructure);

    break; // enable SPI2 clock
```

...
`

**Comment by SytechDesigns** [comment]
@Untitled86 refer to #464 for why your gpio is not set to the correct speed. With the work around in 464 added, gpio can be set to the correct speed and spi will then work fine at full speed ( pclk/2) 96Mhz clock rate.

**Comment by cw2** [comment]
@Untitled86 Also, there is something called I/O compensation cell described in the RM0385:

> By default the I/O compensation cell is not used. However when the I/O output buffer speed
> is configured in 50 MHz or 100 MHz mode, it is recommended to use the compensation cell
> for slew rate control on I/O tf(IO)out)/tr(IO)out commutation to reduce the I/O noise on power
> supply.

**Comment by cw2** [comment]
If anyone has hit problems with UART5, then there is a bug in STM `stm32f746xx.h`:
```C++
#define  RCC_APB1ENR_UART4EN                 0x00080000U
#define  RCC_APB1ENR_UART4EN                 0x00100000U
```
The second symbol should be ...**5**EN.

**Comment by cw2** [comment]
@Untitled86  Now you can unleash the full potential of Visual Studio powerful IDE and your shiny ST-LINK/V2 - #519 :)

**Comment by Untitled86** [comment]
@cw - Thanks!  I can't wait to dive into that.  I've been dealing with more app issues than runtime issues (which is great).  I just wanted to post back with my latest little change.  ...I'm behind on updating this thread, but I plan on posting my full repo soon.

Anyways, I use code based on GHI's OSHW RTC implementation.  It was designed for the F4 and *would* work perfectly with the F7 if BitBanding didn't change/go-away.  If anyone else wants the RTC working in snap, just modify your RTC.cs and change RCC_RTCCLKCmdEnable to the following...

```
            public static void RCC_RTCCLKCmdEnable(UInt32 NewState)
            {
                //I put this here so I could have a simple snibbit.  It would be more consistant to place it with the rest up top. 
                uint RCC_BDCR_RTCEN = 0x00008000; 

                if (NewState != ENABLE)
                {
                    RCC_BDCR.Write(RCC_BDCR.Read() & ~RCC_BDCR_RTCEN);
                }
                else
                {
                    RCC_BDCR.Write(RCC_BDCR.Read() | RCC_BDCR_RTCEN);
                } 
            }
```

Ultimately I want to help improve the RTC integration into the NETMF platform.  I just don't know how I'm supposed to connect my code to the runtime properly.

**Comment by Untitled86** [comment]
@cw - Thanks!  I can't wait to dive into that.  I've been dealing with more app issues than runtime issues (which is great).  I just wanted to post back with my latest little change.  ...I'm behind on updating this thread, but I plan on posting my full repo soon.

Anyways, I use code based on GHI's OSHW RTC implementation.  It was designed for the F4 and _would_ work perfectly with the F7 if BitBanding didn't change/go-away.  If anyone else wants the RTC working in snap, just modify your RTC.cs and change RCC_RTCCLKCmdEnable to the following...

```
            public static void RCC_RTCCLKCmdEnable(UInt32 NewState)
            {
                //I put this here so I could have a simple snibbit.  It would be more consistant to place it with the rest up top. 
                uint RCC_BDCR_RTCEN = 0x00008000; 

                if (NewState != ENABLE)
                {
                    RCC_BDCR.Write(RCC_BDCR.Read() & ~RCC_BDCR_RTCEN);
                }
                else
                {
                    RCC_BDCR.Write(RCC_BDCR.Read() | RCC_BDCR_RTCEN);
                } 
            }
```

Originally the code was as follows, note the BB (BitBanding)...

```
            public static void RCC_RTCCLKCmdEnable(UInt32 NewState)
            {
                SafeToProceed(DeviceID.STM32F4Compatible);

                //Bit 15 RTCEN: RTC clock enable
                new Register(BDCR_RTCEN_BB).Write(NewState);
            }
```

Ultimately I want to help improve the RTC integration into the NETMF platform.  I just don't know how I'm supposed to connect my code to the runtime properly.

**Comment by SytechDesigns** [comment]
@cw2  I have been working on a MF4.3 port for the M7, testing on a Nucleo board. I have LWIP basically working, but there are some timing oddities! DHCP works, then when I ping the device, there is can be a big delay in the ping responses, so that the response for request 1 does not arrive until after request 2 has been sent - so windows ping will report it as timed out - so on 4 requests, I get a 50% failure. 
DHCP is also a bit strange in that it tends to complete before the managed test app has been loaded, and there seems to be a usb connect disconnect sequence there as well. If I run my managed test app which implements an echo test server and run echo test on the pc ( from Pavel Banksi), then I can echo 100 messages ok, but the response time is about 7mSecs, whereas on a F4 port it is 1msec. So it basically works - but there are some timing oddities - have you come across a similar problem with your mf4.4 work, even tho 4.4 is using RTX, lwip is basically the same between 4.3 and 4.4 and I suspect this is more a M4 V M7 type issue.
regards...

**Comment by cw2** [comment]
Hello, unfortunately, I have not had a chance to finish my networking port (due to family reasons), but I've got the initialization stuff working (basic driver for LAN8742 PHY on the Nucleo board), speed negotiation and connection. Now, I don't remember exactly if I am using DHCP, but most likely not (yet) and I know there is some problem further in the networking because the HTTP request fails.

There are a few things that make it a little bit challenging, because I don't have any other (M4) board with working NETMF RTOS networking implementation for reference/comparison; CMSIS RTOS (RTX) does not yet support Cortex-M7, so I had to tweak the existing M4 code and hope it works, and I shot myself in the foot with M7 instruction and data caches.

**Comment by SytechDesigns** [comment]
@cw2 - Hi, I have come to the conclusion that their might be nothing wrong with my LWIP code. It sends and receives etc. but is just a bit slow to respond. This got me thinking about something else I noticed, but filed away to look at later! Tiny booter starts and USB connects, then TinyCLR starts and usb connects - which is different behaviour than on all my other mf work. Normally tiny booter starts and passes over to tinyclr so quickly that windows does not have time to enumerate the usb in tinybooter. I am beginning to suspect that actually the M7 is running slowly - that's why Ethernet seems so 'sluggish'. I have tried turning I Cache on and off, but no real difference, other than it runs quicker ( eth response) with it on - but still slow, and as on your code comment if d cache turned on - things don't really work. ( whats the problem there?)
As a sanity check I have run native code on the Nucleo boards with LwIP and also the MDK Ethernet stack - both work perfectly. I have tried with processor speed at 198Mhz and also 168Mhz  - no difference. My bootstrap code is almost identical to yours - I am beginning to suspect that we are not setting up the processor correctly - any thoughts on this?

**Comment by smaillet-ms** [comment]
@SytechDesigns  DHCP is actually one of the major driving reasons for the move to using an RTOS for LWIP in 4.4. In 4.3 the LWIP stack was updated (apparently fairly late in the process) and the update introduced a subtle bug that comes from how LWIP was adapted to NETMF. In general, like NETMF - LWIP can be built to work single threaded or with a multi-threaded OS. And like NETMF in a single threaded world it wants to be the owner of the single thread. Thus the NETMF port modified the LWIP stack to allow NETMF to be in charge and call into the stack via completions/continuations queues. However, the LWIP stack update include changes that assumed the calls would occur in response to a time to decrement DHCP timeouts among other things. That is the LWIP stack assumed certain calls would only happen at well known periodic intervals. However the, CLR was calling them via the continuations at random intervals including in a loop while there was nothing else to do. Thus the counters were being decremented VERY fast. This triggered DHCP requests to timeout if the server wasn't really fast to respond, furthermore the system would send out renewal requests in much shorter time periods than the server specified. (depending on the system as short as 30s). This caused a tear down of the IP address and breaks all connections. Thus, causing exceptions in .NET code. IF the code wasn't designed to handle network and interface attach/remove then the app would fail.

**Comment by SytechDesigns** [comment]
@smaillet-ms Strangely DHCP is the only thing working well! The eth mac on M7 is compatible with the M4 - so only very minor changes in the driver - just a few constants in the stm32F746.h are back to being int32 rather than int16 defines. Then just different phy - but as on the whole MF uses the 'standard' phy registers, different phys are not really a big issue. Eth on single thread works ( on the m7 mf4.3 port - but its like response to a rx'ed packet is slow. So on a ping request, it can be almost 5 seconds before the response is sent, but the response to the second request is quite quick. Same with the test with tcpip echo server on pc and managed app on device. We get 100+ echos without error, but response time from rx to tx is 160mSecs - on M4 it is <3ms. So I'm not getting data corruption, lost packets etc. just sluggish response.
I need to check if entire MF is running slow, or just eth. I suspect this could be due to the piplining on the M7 making the problems you mentioned on lwip single thread, integrating with clr conflict.I suspect the rest of the MF port is working fine. I need to test more to be sure. But I think that maybe I need to move over to MF4.4 for this project. I will start with normal lwip - but ideally I want to move over to the MDK Ethernet Middle ware stack. This works well on the M7 with RTX and has a BSD interface - so will be easy to integrate into the MF PAL. But what would be helpfull are some pointers into where in the code, the clr is put onto the main RTX thread. The lwip threads are easy to find, but where clr sets up main thread and runs from it is not so obvious. I'm use to RTX projects 'main' calling KernalInitialise, then setting up other threads, and then the main thread just being put into a while forever do nothing loop! So if you could point me in the right direction in the code - it would save some searching!

**Comment by smaillet-ms** [comment]
We just implement main to call the CLR startup, nothing special. The Kernel initialize etc.. from main are not needed in CMSIS-RTX. (The startup libs have their own entry point hooks that takes care of that before calling main in [Main.cpp](https://github.com/NETMF/netmf-interpreter/blob/dev/Solutions/MCBSTM32F400/DeviceCode/TinyHal/main.cpp) ) For the CMSIS part that actually calls KernelInitialize see [RTX_CM_lib.h](https://github.com/ARM-software/CMSIS/blob/master/CMSIS/RTOS/RTX/INC/RTX_CM_lib.h) around line 252

**Comment by JB16** [comment]
@cw2 are you planning on adding support for i2c for STM32F7 anytime soon? Or, could someone help answer some questions to port I2C from F4 to F7 for a c# NETMF project. Looks like the register-set for  I2C is different between F4 and F7, and some of the registers don't seem to have equivalents. Simply put, I'm looking for a STM32F7_I2C_functions.cpp, or atleast want to know what to use instead of CCR as we don't seem to have an equivalent clock register on f7.

**Comment by cw2** [comment]
@JB16 I am not going to add I2C support anytime soon, sorry. F7 has two I2C control registers I2C_CR1 and I2C_CR2 - probably the easiest way is to have a look at STM32 implementation of I2C HAL driver, included in their STM32F7xx_HAL_Driver library (a downloadable part of their STM32CubeMX suite).

**Comment by JB16** [comment]
@cw2 thanks for your quick reply! Do you know of some software that puts out HAL in terms of something similar to STM32Fx_functions.cpp that can be directly used in netmf-interpreter. I already have the application part code-ready, and the netmf-interpreter HAL/PAL drivers are something that I'd prefer getting ready-to-go now.

**Comment by cw2** [comment]
@JB16 Unfortunately, I am not aware of any alternative STM32 HAL implementation that can be used directly in NETMF. Probably the closest to what you'd want is STM32CubeMX that can generate code to initialize hardware peripherals using STM's HAL library - but the interface is rather different, so it has to be modified to match NETMF interface (i.e. by adding necessary wrapper functions). Which is usually too much work, so I only use the STM's code for reference...

**Comment by patrickmealey** [comment]
@cw2 Any chance I could chat with you offline?  My email is in my profile...

Thanks

**Comment by GeorgeStephens** [comment]
@Untitled86 @cw2 Has any progress been made on this/is it possible to post a repo containing what's been done so far? I'm looking doing some work based on the STM32F769 Disco

**Comment by cw2** [comment]
@GeorgeStephens Unfortunately, no progress since then, sorry.

**Comment by Spawn32** [comment]
@Untitled86 
Hi Brien,
Would it be possible for you to share your clock setup you used on the STM32F7 @ 12mhz ? spent 2 days trying to get my usb working, giving up :)

had a look at you github, but couldn't find it.

Would really appreciate it.

**Comment by Spawn32** [comment]
Hmm... is @Untitled86  not here anymore ? :/
Have fixed the USB but cant get SPI going...

**Comment by piwi1263** [comment]
Maybe you can have look here https://github.com/nanoframework/nf-interpreter/tree/develop/targets/CMSIS-OS/ChibiOS/ST_STM32F769I_DISCOVERY for further details on clock, spi etc.
